[APM] Use Zstd compression on Trace Payloads

ichinaski opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

This PR re-introduces the Zstd compression initially implemented on #23806 
The original PR surfaced critical build issues on environments where cgo is not available, therefore it had to be rolled back.

~We now add a new build tag, `zstd`, to determine whether Zstd is available or not, and fall back to gzip compression otherwise.~

To work around OTel limitations, we have now created a new `trace/compression` component, with 2 implementations:
* `impl-gzip`: existing gzip compression algorithm, it represents no change compared to current compression logic.
* `impl-zstd`: Zstandard compression, requires cgo.

This component also defines 3 new Go modules, in the `/def`,  `/impl-gzip`, and `/impl-zstd`. This is requested by OTel, and goes in accordance with the [Components documentation](https://datadoghq.dev/datadog-agent/components/creating-components/#go-module)

To enable Zstd, `zstd-encoding` configuration setting must be passed (intentionally undocumented). When this setting is pased, trace-agent will use `zstd`. OTel usages are configured to always use `gzip`, and never import `impl-zstd` (this preventing build issues when cgo is not available).

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

From the original PR #23806 

> Investigations showed that we can have better compression of our payloads without performance impacts. This can lead to significant data transfer cost savings on our end, and cost savings for our customers as well.

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

From the original PR #23806 
> More information on the initial investigation in this [doc](https://docs.google.com/document/d/1HiMsdz8UDQEI-2Ry-lycCfvuMuCow_Scon8JWhCi9qc/edit)

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

**Scenario 1**:

Without changes, sending traces should still work out of the box. Encoding type should be `gzip`.

**Scenario 2**:

Add `zstd-encoding` in DD_APM_FEATURES
Sending traces work. Encoding type should be `zstd`
CPU usage shouldn't be higher than in the 1st scenario.

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

ajgajg1134 committed:
[APM] Reintroduce zstd compression commits

ichinaski committed:
[APM] Enable Zstd by default on Trace payloads

Zstd requires cgo, which is not available on all environments. This PR
adds a new build tag "zstd", that allows TraceWriter to use Zstd
encoding if it's available, or fall back to gzip if not.

An optional `disable-zstd-encoding` config setting allows to manually
disable zstd, and use gzip instead.

ichinaski committed:
Add release notes

dineshg13 commented on pkg/trace/writer/zstd.go:
```diff
@@ -0,0 +1,24 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build zstd
+
+package writer
+
```
Build tags don't work with OTel Collector. Please make zstd writer an interface than can have noop implementation.

ichinaski replied:
If no build tag is passed, `zstd` will not be available during the build process. (and we will fall back to gzip). How is this incompatible with OTel Collector?

ichinaski committed:
[APM] Disable Zstd enconding by default

To have more control over the rollout and be able to gradually test this
feature, the new enconding will be disabled by default (as originally
planned).

ichinaski committed:
[APM] Add zstd build tag to TRACE_AGENT

ichinaski committed:
[APM] Update deps and fix out-of-sync  errors

ichinaski committed:
[APM] Format built_tags.py

ichinaski committed:
Merge branch 'main' into ichinaski/zstd

ichinaski committed:
[APM] Extract compression into standalone module (#27045)

GustavoCaso commented on tasks/build_tags.py:
```diff
@@ -141,7 +141,17 @@
+    "otlp",
+    "netcgo",
+    "podman",
+    "zstd",
```
Is this tag still relevant?

ichinaski replied:
It is not, good catch!

ichinaski committed:
update components README

GustavoCaso commented on pkg/serverless/trace/trace.go:
```diff
@@ -99,7 +102,13 @@ func (s *ServerlessTraceAgent) Start(enabled bool, loadConfig Load, lambdaSpanCh
 			tc.SynchronousFlushing = true
-			s.ta = agent.NewAgent(context, tc, telemetry.NewNoopCollector(), &statsd.NoOpClient{})
+			var compressor compression.Component
+			if tc.HasFeature("zstd-encoding") {
+				compressor = zstd.NewComponent()
+			} else {
+				compressor = gzip.NewComponent()
+			}
```
is this something we want to allow customers to tweak in the serverless environment. Would it be easier to use the same option always?

ichinaski replied:
This is actually an undocumented setting, that is meant to be hidden from the customer. We want to silently provide this option for us to gradually enable zstd on org2 prod clusters, and ensure no issue arises. If all goes well, the goal is to flip the setting in the next release, and use zstd by default in all cases.

ichinaski committed:
Merge branch 'main' into ichinaski/zstd

ichinaski committed:
rm unnecessary build tag

ichinaski committed:
fix otel tests

dineshg13 commented on comp/trace/compression/compression.go:
```diff
@@ -0,0 +1,7 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package compression provides a compression implementation based on the configuration or available build tags.
```
remove `available build tags` ?

dineshg13 commented on comp/trace/agent/impl/agent.go:
```diff
@@ -116,11 +119,18 @@ func NewAgent(deps dependencies) (traceagent.Component, error) {
 		return nil, err
 	}
 	setupShutdown(ctx, deps.Shutdowner, statsdCl)
+	var compressor compression.Component
+	if tracecfg.HasFeature("zstd-encoding") {
+		compressor = zstd.NewComponent()
+	} else {
+		compressor = gzip.NewComponent()
+	}
```
Can we add compressor as fx deps and set the compressor in trace agent cmd ?

ajgajg1134 committed:
Merge branch 'main' into ichinaski/zstd

ichinaski committed:
Set compressor as fx dependency

ichinaski committed:
fix lint errors

dineshg13 reviewed (approved):
LGTM. Thanks for making all the changes. QQ: Do you think we should use zstd for OTel Agent ? OTel Agent isn't used by any customers and we are still in testing phase.

ichinaski commented:
> LGTM. Thanks for making all the changes. QQ: Do you think we should use zstd for OTel Agent ? OTel Agent isn't used by any customers and we are still in testing phase.

I see no reason not to use the same approach as the trace-agent for now: if `zstd-encoding` feature is available, we use zstd, otherwise and by default, gzip.

If you think this makes sense I can add it to the otel-agent commands.

dineshg13 commented:
Lets merge this in. There is no hurry to make the changes. We can default to zstd in follow up PR

GustavoCaso commented on comp/trace/bundle_test.go:
```diff
@@ -39,6 +41,7 @@ func TestBundleDependencies(t *testing.T) {
 		statsd.Module(),
 		fx.Provide(func(cfg config.Component) telemetry.TelemetryCollector { return telemetry.NewCollector(cfg.Object()) }),
 		fx.Supply(tagger.NewFakeTaggerParams()),
+		fx.Provide(func(cfg config.Component) compression.Component { return gzip.NewComponent() }),
```
We can remove the `cfg config.Component` from the function arguments as we are not using it to decide the compression component 😄

GustavoCaso commented on cmd/otel-agent/subcommands/run/command.go:
```diff
@@ -165,6 +167,9 @@ func runOTelAgentCommand(ctx context.Context, params *subcommands.GlobalParams,
 		fx.Provide(func(cfg traceconfig.Component) telemetry.TelemetryCollector {
 			return telemetry.NewCollector(cfg.Object())
 		}),
+		fx.Provide(func(cfg traceconfig.Component) tracecompression.Component {
+			return gzip.NewComponent()
+		}),
```
If I'm not mistake we should provide the fx version of each component. that way we could simply do 

```go
gzip "github.com/DataDog/datadog-agent/comp/trace/compression/fx-gzip/fx"
...


gzip.Module(),
...

```

You can see in the documentation and example of the different fx related folder 
https://datadoghq.dev/datadog-agent/components/creating-components/#the-fx-folders-compcompressionfx-zstdfxgo

GustavoCaso commented on cmd/trace-agent/subcommands/run/command.go:
```diff
@@ -113,6 +116,12 @@ func runTraceAgentProcess(ctx context.Context, cliParams *Params, defaultConfPat
 			MemProfile:  cliParams.MemProfile,
 			PIDFilePath: cliParams.PIDFilePath,
 		}),
+		fx.Provide(func(cfg config.Component) compression.Component {
+			if cfg.Object().HasFeature("zstd-encoding") {
+				return zstd.NewComponent()
+			}
+			return gzip.NewComponent()
+		}),
```
Discussed with @ichinaski over Slack.

For this case we can not use `gzip.Module()` or `zstd.Module()` as Fx won't be able to resolve the type. For now we agree to leave as it is.

ichinaski committed:
add fx folders to compression comp

ichinaski committed:
mend

ichinaski committed:
mend

ichinaski committed:
mend

GustavoCaso commented on comp/trace/compression/go.mod:
```diff
@@ -0,0 +1,27 @@
+module github.com/DataDog/datadog-agent/comp/trace/compression
```
Should we move the Go module from the top level directory? Just create the module if the implementation we want to share.

GustavoCaso commented on comp/otelcol/otlp/components/statsprocessor/go.sum:
```diff
@@ -8,6 +8,8 @@ github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/attributes v0.16.1 h1:ZI8u3
 github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/attributes v0.16.1/go.mod h1:dvIWN9pA2zWNTw5rhDWZgzZnhcfpH++d+8d1SWW6xkY=
 github.com/DataDog/sketches-go v1.4.2 h1:gppNudE9d19cQ98RYABOetxIhpTCl4m7CnbRZjvVA/o=
 github.com/DataDog/sketches-go v1.4.2/go.mod h1:xJIXldczJyyjnbDop7ZZcLxJdV3+7Kra7H1KMgpgkLk=
+github.com/DataDog/zstd v1.5.5 h1:oWf5W7GtOLgp6bciQYDmhHHjdhYkALu6S/5Ni9ZgSvQ=
+github.com/DataDog/zstd v1.5.5/go.mod h1:g4AWEaM3yOg3HYfnJ3YIawPnVdXJh9QME85blwSAmyw=
```
is this expected @dineshg13 ?

ichinaski committed:
update test file with fx module

ichinaski committed:
Split module into def,impl-gzip,impl-zstd

These are the modules needed by OTel, not the one at the root

ichinaski committed:
update module dependencies

ichinaski committed:
format python file

ichinaski committed:
Revert unintended go version bump

StephenWakely commented on comp/trace/compression/impl-zstd/zstd.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+func (c *compressor) NewWriter(w io.Writer) (io.WriteCloser, error) {
+	return zstd.NewWriterLevel(w, zstd.BestSpeed), nil
```
I'm wondering if we want to standardise on the compression level the agent sets. Currently Metrics sets it by default to `DefaultCompression` (5). But this can be configured by the user. I was going to do similar for logs.

Is there a reason `BestSpeed` has been chosen here? Maybe we should change metrics to do the same level.

ichinaski replied:
`BestSpeed` provided the highest throughput on the trace-agent, with just a slightly lower compression ratio than the default one. See the linked document for the full analysis: https://docs.google.com/document/d/1HiMsdz8UDQEI-2Ry-lycCfvuMuCow_Scon8JWhCi9qc/edit#heading=h.ei4su89bdi9f

StephenWakely replied:
That's quite a boost!

ichinaski committed:
Add missing licenses

ichinaski committed:
Merge branch 'main' into ichinaski/zstd

ichinaski committed:
Remove commented code

We will work on this sync.Pool implementation on a separate PR

GustavoCaso commented on comp/otelcol/collector-contrib/impl/go.mod:
```diff
@@ -765,6 +767,12 @@ replace github.com/DataDog/datadog-agent/comp/otelcol/otlp/components/metricscli
+
+replace github.com/DataDog/datadog-agent/comp/trace/compression/impl-gzip => ../../../trace/compression/impl-gzip
+
+replace github.com/DataDog/datadog-agent/comp/trace/compression/impl-zstd => ../../../trace/compression/impl-zstd
```
We can remove this replace as we no longer use zstd implementation in OTel

ichinaski replied:
all OTel modules that depend on `/pkg/trace/` end up requiring all these implementations, at least in their go.mod, since `pkg/trace/writer` is still directly using them for testing purposes.

We could decouple the component from the trace-writer tests and somewhat simplify these dependencies, but that would be a separate PR.

GustavoCaso commented on comp/otelcol/otlp/components/exporter/datadogexporter/go.mod:
```diff
@@ -24,6 +24,9 @@ replace (
 	github.com/DataDog/datadog-agent/comp/trace/agent/def => ../../../../../../comp/trace/agent/def
+	github.com/DataDog/datadog-agent/comp/trace/compression/def => ../../../../../../comp/trace/compression/def
+	github.com/DataDog/datadog-agent/comp/trace/compression/impl-gzip => ../../../../../../comp/trace/compression/impl-gzip
+	github.com/DataDog/datadog-agent/comp/trace/compression/impl-zstd => ../../../../../../comp/trace/compression/impl-zstd
```
We can remove this replace as we no longer use zstd implementation in OTel

GustavoCaso commented on comp/otelcol/otlp/components/statsprocessor/go.mod:
```diff
@@ -4,6 +4,9 @@ go 1.21.0
 	github.com/DataDog/datadog-agent/comp/otelcol/otlp/components/metricsclient => ../metricsclient
+	github.com/DataDog/datadog-agent/comp/trace/compression/def => ../../../../../comp/trace/compression/def
+	github.com/DataDog/datadog-agent/comp/trace/compression/impl-gzip => ../../../../../comp/trace/compression/impl-gzip
+	github.com/DataDog/datadog-agent/comp/trace/compression/impl-zstd => ../../../../../comp/trace/compression/impl-zstd
```
We can remove this replace as we no longer use zstd implementation in OTel

GustavoCaso commented on pkg/trace/stats/oteltest/go.mod:
```diff
@@ -89,6 +91,9 @@ require (
 replace (
+	github.com/DataDog/datadog-agent/comp/trace/compression/def => ../../../../comp/trace/compression/def
+	github.com/DataDog/datadog-agent/comp/trace/compression/impl-gzip => ../../../../comp/trace/compression/impl-gzip
+	github.com/DataDog/datadog-agent/comp/trace/compression/impl-zstd => ../../../../comp/trace/compression/impl-zstd
```
We can remove this replace as we no longer use zstd implementation in OTel

GustavoCaso commented on tasks/modules.py:
```diff
@@ -182,6 +182,13 @@ def dependency_path(self, agent_version):
+        "comp/trace/compression/impl-gzip", independent=True, used_by_otel=True
+    ),
+    "comp/trace/compression/impl-zstd": GoModule(
+        "comp/trace/compression/impl-zstd", independent=True, used_by_otel=True
```
I think this is not by Otel 😄

GustavoCaso commented on test/otel/go.mod.fake:
```diff
@@ -18,6 +18,9 @@ replace (
 	github.com/DataDog/datadog-agent/comp/serializer/compression => ./../../comp/serializer/compression
+	github.com/DataDog/datadog-agent/comp/trace/compression/def => ./../../comp/trace/compression/def
+	github.com/DataDog/datadog-agent/comp/trace/compression/impl-gzip => ./../../comp/trace/compression/impl-gzip
+	github.com/DataDog/datadog-agent/comp/trace/compression/impl-zstd => ./../../comp/trace/compression/impl-zstd
```
is zstd relevant here?

ichinaski commented:
/merge
